### PR TITLE
Feature/#258 search query and id list

### PR DESCRIPTION
**Pull request by alefnula**
Implemented a new parser for classic api query using `lark-parser` (it's faster and more flexible and has nicer syntax than ply).

Rewrote all the tests and added tests for ElasticSearch query generation.

I still have to test this a little bit, but as far as I did until now everything works as it should be.

### CODE REVIEW

File: search/domain/base.py [hunk 1]
```diff
@@ -195,22 +200,38 @@ class Field(Enum):
...
-   phrase: Phrase = (('au', 'del_maestro'), 'AND', ('ti', 'checkerboard'))
+   phrase: Phrase = (
+       Operator.AND,
+       Term(Field.Author, "del_maestro"),
+       Term(Field.Title, "checkerboard")
+    )
 
    # Disjunct query with an unary not.
-   phrase = (('au', 'del_maestro'), 'OR', ('ANDNOT', ('ti', 'checkerboard')))
+   phrase = (
```
Comment (mhl10) [inline search/domain/base.py]: I think this is much easier to read than before 👍

File: search/domain/classic_api/query_parser.py [hunk 2]
```diff
@@ -1,208 +1,126 @@
 """
 Utility module for Classic API Query parsing.
 
-We use a recursive descent parser, implemented via :func:`parse_classic_query`.
+Uses lark-parser (EBNF parser).
```
Comment (mhl10) [inline search/domain/classic_api/query_parser.py]:
```suggestion
Uses lark-parser (EBNF parser) [1].
[1]: https://github.com/lark-parser/lark/blob/master/README.md
```

File: search/domain/classic_api/query_parser.py [hunk 3]
```diff
@@ -1,208 +1,126 @@
...
+from werkzeug.exceptions import BadRequest
 
-from typing import List, Optional, Union
+from search.domain.base import Operator, Field, Term, Phrase
 
-from werkzeug.exceptions import BadRequest
 
-from search.domain.base import Phrase, Operator, Field, Term
+class QueryTransformer(Transformer):
+    def string(self, tokens):
```
Comment (mhl10) [inline search/domain/classic_api/query_parser.py]: A class docstring would be helpful here

File: search/domain/classic_api/query_parser.py [hunk 4]
```diff
@@ -1,208 +1,126 @@
...
+        return None
 
+    def query(self, tokens):
+        return tokens[0]
 
-def _parse_field_query(field_part: str) -> Term:
-    field_name, value = field_part.split(":", 1)
 
-    # Cast field to Field enum.
+QUERY_PARSER = Lark(
```
Comment (mhl10) [inline search/domain/classic_api/query_parser.py]: super nice!

File: search/services/index/api_classic/classic_search.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    # Initialize query.
+    if query.phrase:
+        dsl_query = query_builder(query.phrase)
+    else:
+        dsl_query = Q()
+
+    # Filter id_list if necessary.
+    if query.id_list:
+        # Separate versioned and unversioned papers.
+        paper_ids = [id for id in query.id_list if "v" not in id]
```
Comment (mhl10) [inline search/services/index/api_classic/classic_search.py]: This unfortunately won't work for a subset of the old-style arXiv identifiers (e.g. `solv-int/9704004`). Something like this should be ok:
```
import re
r1 = re.compile(".*v[1-9]\d*$")
paper_ids = []
paper_id_vs = []
for id in query.id_list:
    paper_id_vs.append(id) if r1.search(id) else paper_ids.append(id)

**Comment by alefnula** [comment]
I've spent some time comparing the EBNF parsers since I got a little bit stuck with PLY (simple things were simple but then I got stuck with some of more complex stuff). And then I found LARK which is more actively developed, has more github stars, is pretty much faster and uses a lot less memory. I really don't know why I haven't heard of it earlier, but doing it in LARK was a breeze. So I think it's a win win situation.

There is only one thing that I want to fix. But I need a suggestion. What should an empty string parse to? `all:""`?

**Comment by mhl10** [comment]
> There is only one thing that I want to fix. But I need a suggestion. What should an empty string parse to? `all:""`?

If `search_query` [is empty](http://export.arxiv.org/api/query?search_query) or any of the [field terms are empty](http://export.arxiv.org/api/query?search_query=au:) I would just short-circuit the request to the search backend and return something like the legacy version does, with no entries and `200` status. It's a bit odd, but would be compatible I suppose.

**Comment by alefnula** [comment]
OK will do that.

File: search/services/index/api_classic/classic_search.py (hunk 5 above)
Comment (alefnula) [reply search/services/index/api_classic/classic_search.py]: Fixed.

File: search/domain/classic_api/query_parser.py (hunk 3 above)
Comment (alefnula) [reply search/domain/classic_api/query_parser.py]: Added.

**Comment by alefnula** [comment]
I just have to add the edge cases where no query string is provided and then this pull request is ready.

**Pull request merged**